### ISSUE TITLE
Console Aplication (C#) - Memory Leak

### ISSUE THREAD

**Issue by Cacuci** [dotnet/runtime#45315]
I have a program "Console Aplication" install in Windows that execute every time. Initial consume 32 MB RAM, but grow infinitely. My program execute connection with DB oracle through provider "Oracle.ManagedDataAccess". To Diagnostics what was happening, i executed o PerfView, the result show high consume/execute in namespace System.Diagnostics and Microsoft.Win32 that GB (Garbage Colletion) can´t solve.
  -->

### Configuration

* Which version of .NET is the code running on? 
  NET 4.8

* What OS and version, and what distro if applicable? 
  Windows Server 2008 and Windows Server 2016

* What is the architecture (x64, x86, ARM, ARM64)? 
  Architecture x64,

* Do you know whether it is specific to that configuration? 
   No, I don´t.

* If you're using Blazor, which web browser(s) do you see this issue in? 
  Not applicable

### Other information

![imagem1](https://user-images.githubusercontent.com/8560576/100550466-cd6e4980-3258-11eb-8695-49c4283bb084.jpg)

![imagem2](https://user-images.githubusercontent.com/8560576/100550469-d3642a80-3258-11eb-843a-2f836dd51866.jpg)

**Comment by GrabYourPitchforks** [comment]
Note to area owners: this report is about a Full Framework issue.

**Comment by eiriktsarpalis** [comment]
If I get this correctly, is it just the two static variables `System.Diagnostics.NtProcessInfoHelper.CachedBuffer` and `Microsoft.Win32.Registry.CurrentUser` that grow indefinitely?

Would you be able to share a minimal reproduction of the issue?

**Comment by eiriktsarpalis** [comment]
Hi @Cacuci, any updates on the questions I posted earlier?

**Comment by eiriktsarpalis** [comment]
Assuming this issue is related to [ProcessManager.Win32](https://github.com/dotnet/runtime/blob/839b1343fbad61526fedcaa722f1ae48b298c543/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs#L246), I believe we could improve this implementation, memory leak or no memory leak. @adamsitnik suggested using `ArrayPool` and dealing with alignment manually.

**Comment by Cacuci** [comment]
Thank you @eiriktsarpalis, this solution probably resolve the problem. My program work with LOH (Large Object Heap), why my Queries in oracle return objects (XML) bigger than 85 KB.

**Comment by eiriktsarpalis** [comment]
@Cacuci I would say it's mostly an unrelated enhancement. We would still need some form of minimal reproduction to take a better look at your issue.

> If I get this correctly, is it just the two static variables System.Diagnostics.NtProcessInfoHelper.CachedBuffer and Microsoft.Win32.Registry.CurrentUser that grow indefinitely?

Would you be able to confirm or disconfirm this? Thanks

**Comment by adamsitnik** [comment]
Hi @Cacuci 

To be able to identify the source of the memory leak you need to perform a full memory leak analysis.

Before you do that, please watch these three tutorial videos:

1. [PerfView Tutorial 9 - .NET Memory Investigation: Basics of GC Heap Snapshots](https://channel9.msdn.com/Series/PerfView-Tutorial/PerfView-Tutorial-9-NET-Memory-Investigation-Basics-of-GC-Heap-Snapshots)
1. [Tutorial 10 - Investigating .NET Heap Memory Leaks :Part1 Collecting the dat](https://channel9.msdn.com/Series/PerfView-Tutorial/Tutorial-10-Investigating-NET-Heap-Memory-Leaks-Part1-Collecting-the-data)
1. [Tutorial 11 - Investigating .NET Heap Memory Leaks :Part2 Analyzing the data](https://channel9.msdn.com/Series/PerfView-Tutorial/Tutorial-11-Investigating-NET-Heap-Memory-Leaks-Part2-Analyzing-the-data)

After that, you need to:

1. Start your app, perform the warmup (like invoke the web service once) to make sure that the code gets JITed and all static caches get populated
2. Force GC from memory profiler and capture a memory dump (1)
3. Perform the work that you believe leaks the memory (like close and open a Window or perform a single web request)
4. Force GC from memory profiler and capture a memory dump (2)
5. Use Memory Profiler to diff both trace files so you get only the objects that were born between first and second snapshot but did not die after forcing GC when taking second snapshot. The diff should contain the leaks (things that should have been released but were not). Use the profiler to find out what keeps the object alive and modify your code to clean these references (by Disposing resources, unregistering from events, removing static variables, setting things to null, etc)


Once you do that please get back to us with what you have found. If there is a leak in .NET and it's reproducible in .NET Core (it's very unlikely) we are going to fix it.

Thanks,
Adam

**Comment by danmosemsft** [comment]
Do we need a separate issue to replace static with  ArrayPool? That seems worth doing.

**Comment by adamsitnik** [comment]
@danmosemsft I am already working on a PR : https://github.com/dotnet/runtime/pull/45690

**Comment by eiriktsarpalis** [comment]
@adamsitnik has already created PR #45690 so perhaps that might suffice?

**Comment by Cacuci** [comment]
> @Cacuci I would say it's mostly an unrelated enhancement. We would still need some form of minimal reproduction to take a better look at your issue.
> 
> > If I get this correctly, is it just the two static variables System.Diagnostics.NtProcessInfoHelper.CachedBuffer and Microsoft.Win32.Registry.CurrentUser that grow indefinitely?
> 
> Would you be able to confirm or disconfirm this? Thanks

**Comment by Cacuci** [comment]
> @Cacuci I would say it's mostly an unrelated enhancement. We would still need some form of minimal reproduction to take a better look at your issue.
> 
> > If I get this correctly, is it just the two static variables System.Diagnostics.NtProcessInfoHelper.CachedBuffer and Microsoft.Win32.Registry.CurrentUser that grow indefinitely?
> 
> Would you be able to confirm or disconfirm this? Thanks

Confirm @eiriktsarpalis. Two static variables System.Diagnostics.NtProcessInfoHelper.CachedBuffer and Microsoft.Win32.Registry.CurrentUser that grow indefinitely

**Comment by Cacuci** [comment]
Sorry, i closed unduly.

**Comment by adamsitnik** [comment]
@Cacuci could you please follow the steps I've described [above](https://github.com/dotnet/runtime/issues/45315#issuecomment-740714288) and share the dump files diff, not just a single dump file? What you have shared shows all the objects that live in the memory, not the actual memory leaks

**Comment by Cacuci** [comment]
Yes @adamsitnik, i will reproduce your sugestion and post the results. Thank you!

**Comment by Cacuci** [comment]
@adamsitnik Follow imagens, dump, dump2, dump (diff) and .dump files in .zip.

![Dumps](https://user-images.githubusercontent.com/8560576/101708994-519caa00-3a6d-11eb-8065-088dd9805348.PNG)

Dumps (Diff)

![dump(Diff)](https://user-images.githubusercontent.com/8560576/101709048-76911d00-3a6d-11eb-864a-1a60962af227.PNG)

Dump files: [DUMPS.zip](https://github.com/dotnet/runtime/files/5669639/DUMPS.zip)

**Comment by adamsitnik** [comment]
From the dumps that you have provided, we can see that in the first dump there was:

527791 bytes allocated by 4261 objects

in the second:

511337 bytes allocated by 3052 objects

So... the **memory usage did not increase but has decreased instead** (and the diff window shows a negative `Totals Metric`).

To be 100% sure there is no leak I've opened the dumps and diffed them:

![obraz](https://user-images.githubusercontent.com/6011991/102069967-c3bf1700-3dfe-11eb-95ea-c311a8947bc8.png)

There are only a few rows with a positive value in the exclusive column. I've checked only a few of them, and they were all rooted by static variables:

![obraz](https://user-images.githubusercontent.com/6011991/102071037-3a104900-3e00-11eb-834e-e4d017c9c41b.png)

![obraz](https://user-images.githubusercontent.com/6011991/102071074-472d3800-3e00-11eb-80e5-07f4d24fe182.png)

![obraz](https://user-images.githubusercontent.com/6011991/102071118-557b5400-3e00-11eb-9228-6ba43eb30ba9.png)

Have you performed the first step from my instructions?

> perform the warmup (like invoke the web service once) to make sure that the code gets JITed and all static caches get populated

`System.Runtime.Serialization.CodeGenerator.stringLength` is being initialized only once both in .NET and .NET Core: 

https://referencesource.microsoft.com/#System.Runtime.Serialization/System/Runtime/Serialization/CodeGenerator.cs,1906

https://github.com/dotnet/runtime/blob/e2099f62f1ee28241f857edaf0ef7420018d11d8/src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/CodeGenerator.cs#L1428

@Cacuci Please perform a proper warmup before taking the first dump, so this and other static variables do not show up in the diff, and share the results of your analysis.

> Initial consume 32 MB RAM, but grow infinitely

If you don't find any leaks, the growth might be caused by the fact that GC has assumed that you have enough memory available and kept allocating more rather than collecting the garbage (it's cheaper). You can use `Performance Monitor` (ships with Windows by default AFAIR) to see if GC performs the cleanups over time:

![obraz](https://user-images.githubusercontent.com/6011991/102072874-c0c62580-3e02-11eb-9ec0-8b0cc9a28f19.png)

If you want to lear more about .NET Memory Performance Analysis please read the following amazing doc written by @Maoni0 : https://github.com/Maoni0/mem-doc/blob/master/doc/.NETMemoryPerformanceAnalysis.md

**Comment by Cacuci** [comment]
@adamsitnik I give up did attempts investigate what's is happing and i update my program to .Net5 and problem did resolved. Previously, with .Net 4.8 a memory did grow infinitely, until the program stop and computer run out of memory, now with .Net5 consumes between 100 MB and 120MB, nothing more that this.

**Comment by adamsitnik** [comment]
@Cacuci thank you for your cooperation and upgrading to .NET 5. Since there is nothing actionable (no memory leak found), I've closed the issue.
